### PR TITLE
Host Ir Lowering: foundation

### PR DESCRIPTION
**Pull request by samnordmann**
# What
Refactor Host Ir lowering. https://jirasw.nvidia.com/browse/NVFUSER-105
Dependent on:
- [x] https://github.com/NVIDIA/Fuser/pull/3524

# Why
- cleaner code
- prepare for integration of lowering of sharded matmul with pipelined algorithms, https://jirasw.nvidia.com/browse/NVFUSER-106

# How
- move files `multidevice/lower_communication` to `host_ir/lower`
- Move most of `MultiDeviceExecutor`'s constructor logic to `host_ir/lower`

 I divided in incremental commits to ease the review. The relevant commits start at [move HostIr lowering to host ir folder and special class](https://github.com/NVIDIA/Fuser/pull/3535/commits/25f09e980fb5096277072bfee7ac132401a87f7e), the commits before that belong to https://github.com/NVIDIA/Fuser/pull/3524

### CODE REVIEW

**Comment by samnordmann** [comment]
!test --diff

**Comment by samnordmann** [comment]
!test

**Comment by samnordmann** [comment]
@wujingyue the CI failures seen on the previous commits seem to be caused by a too small tolerance rate on single device test. It doesn't seem related to the present PR
https://gitlab-master.nvidia.com/dl/pytorch/fuser-gh-mirror/-/jobs/127287825

File: csrc/host_ir/lower.cpp [hunk 1]
```diff
@@ -296,7 +302,10 @@ std::vector<Communication*> lowerCommunication(Expr* c) {
   return comms;
 }
 
-bool isLowerableToCommunication(Expr* expr) {
+bool HostIrLower::canLower(Expr* expr) {
+  if (!isResharding(expr)) {
+    return true;
+  }
```
Comment (wujingyue) [inline csrc/host_ir/lower.cpp:308]: This apparently the behavior and therefore not a pure refactoring. I'm not requesting a breakdown, but can you teach me why this check didn't exist before and/or why we need it now?

File: csrc/host_ir/lower.cpp [hunk 2]
```diff
@@ -325,4 +334,100 @@ bool isLowerableToCommunication(Expr* expr) {
   }
 }
 
+std::unique_ptr<hir::HostIrContainer> HostIrLower::lower(
```
Comment (wujingyue) [inline csrc/host_ir/lower.cpp:337]: Can you confirm it's a straight copy, so I can skip reviewing the content?

File: csrc/multidevice/utils.cpp [hunk 3]
```diff
@@ -555,12 +555,12 @@ std::set<DeviceIdxType> involvedDevices(Expr* expr) {
...
-          tv->hasDeviceMesh(),
-          "the TensorView has no device mesh: ",
-          tv->toString());
-      auto& mesh = tv->getDeviceMesh().vector();
-      std::copy(mesh.begin(), mesh.end(), std::inserter(ret, ret.end()));
+      if (tv->hasDeviceMesh()) {
+        auto& mesh = tv->getDeviceMesh().vector();
+        std::copy(mesh.begin(), mesh.end(), std::inserter(ret, ret.end()));
+      } else {
+        ret.insert(0);
```
Comment (wujingyue) [inline csrc/multidevice/utils.cpp:562]: Why is this change needed?

File: tests/cpp/test_resharding.cpp [hunk 4]
```diff
@@ -40,8 +40,7 @@ class ReshardingTest : public NVFuserFixtureParamTest<ReshardingTestParams> {
     // FusionExecutorCache, simplify validation by using
     // FusionExecutorCache::getMostRecentKernelRuntime()->fusionSegments()->groups().
     for (auto expr : fusion_->exprs()) {
-      EXPECT_TRUE(!isResharding(expr) || isLowerableToCommunication(expr))
-          << "on expr=" << expr;
+      EXPECT_TRUE(HostIrLower::canLower(expr)) << "on expr=" << expr;
```
Comment (wujingyue) [inline tests/cpp/test_resharding.cpp:43]:
```suggestion
      EXPECT_TRUE(HostIrLower::canLower(expr)) << "on expr: " << expr;
```
Nit: it's slightly more readable to use `:` because `expr->toString()` likely contains `=`.

**Review by wujingyue — COMMENTED** [review]
LGTM overall

File: csrc/multidevice/utils.cpp (hunk 3 above)
Comment (samnordmann) [reply csrc/multidevice/utils.cpp:562]: `involvedDevices` is now called from `HostIrEvaluator` which can take-in single device fusion where TVs don't have device mesh. In a single device Fusion, it is natural that this function returns `device 0`

File: csrc/host_ir/lower.cpp (hunk 2 above)
Comment (samnordmann) [reply csrc/host_ir/lower.cpp:337]: it is a straight copy

File: csrc/host_ir/lower.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/host_ir/lower.cpp:308]: I don't think there is any change of behavior in the current places `canLower` is called. However you are right that `isLowerableToCommunication` became `HostIrLower::canLower` and that those methods don't return the same on a non-resharding expr.

The motivation is that Host Ir Lowering will no longer deal with pure communication, but also with compute. Any non-resharding op can be lowered to a `PostOnStream(HostUnit(Fusion))`, or into a Host Op like `MatmulOp` (which will be expr-evaluated). So it is natural that `HostIrLower::canLower` returns `true` on a non-resharding op.

I btw recall that in the coming PR, we will add lowering of distributed matmuls, which embeds both compute and comm.

As a side note, `canLower` should probably eventually be used in the Host segmenter, to know if a segment is valid (the same `canSchedule` is used).

**Review by samnordmann — COMMENTED** [review]
Let me know if you have further questions or comments

File: tests/cpp/test_multidevice_pipeline.cpp [hunk 5]
```diff
@@ -126,7 +125,9 @@ void PipelineTest::executeAndValidate(bool validate_with_prescribed_values) {
   }
 
   runtime = std::make_unique<MultiDeviceExecutor>(
-      std::move(fusion), *communicator_, host_ir_executor_params);
+      std::make_unique<Fusion>(*fusion),
```
Comment (wujingyue) [inline tests/cpp/test_multidevice_pipeline.cpp:128]: Why this change? `fusion` doesn't seem to be used after.

**Review by wujingyue — APPROVED** [review]

File: tests/cpp/test_multidevice_pipeline.cpp (hunk 5 above)
Comment (samnordmann) [reply tests/cpp/test_multidevice_pipeline.cpp:128]: you're right, let me revert it.

**Comment by samnordmann** [comment]
!test
